src/graphics/reproject-texture.js

```diff
@@ -40,7 +44,322 @@ function getProjectionName(projection) {
         case TEXTUREPROJECTION_EQUIRECT: return "Equirect";
         case TEXTUREPROJECTION_OCTAHEDRAL: return "Octahedral";
     }
+};
+
+// pack a 32bit floating point value into RGBA8
+const packFloat32ToRGBA8 = (value, array, offset) => {
+    if (value <= 0) {
+        array[offset + 0] = 0;
+        array[offset + 1] = 0;
+        array[offset + 2] = 0;
+        array[offset + 3] = 0;
+    } else if (value >= 1.0) {
+        array[offset + 0] = 255;
+        array[offset + 1] = 0;
+        array[offset + 2] = 0;
+        array[offset + 3] = 0;
+    } else {
+        let encX = (1 * value) % 1;
+        let encY = (255 * value) % 1;
+        let encZ = (65025 * value) % 1;
+        const encW = (16581375.0 * value) % 1;
+
+        encX -= encY / 255;
+        encY -= encZ / 255;
+        encZ -= encW / 255;
+
+        array[offset + 0] = Math.min(255, Math.floor(encX * 256));
+        array[offset + 1] = Math.min(255, Math.floor(encY * 256));
+        array[offset + 2] = Math.min(255, Math.floor(encZ * 256));
+        array[offset + 3] = Math.min(255, Math.floor(encW * 256));
+    }
+};
+
+// generate a vector on the hemisphere with constant distribution.
+// function kept because it's useful for debugging
+// vec3 hemisphereSampleUniform(vec2 uv) {
+//     float phi = uv.y * 2.0 * PI;
+//     float cosTheta = 1.0 - uv.x;
+//     float sinTheta = sqrt(1.0 - cosTheta * cosTheta);
+//     return vec3(cos(phi) * sinTheta, sin(phi) * sinTheta, cosTheta);
+// }
+
+// generate a vector on the hemisphere with phong reflection distribution
+const hemisphereSamplePhong = (dstVec, x, y, specularPower) => {
```

Comment:
I wonder if these should go to random.js?

Reply:
also, if there are some references / papers these are based on, it'd be great to list.

Reply:
I would prefer to consider this in followup PR?